src/middlewared/middlewared/plugins/asigra.py

```diff
@@ -0,0 +1,358 @@
+        asigra_pg_hba_conf = os.path.join(asigra_postgresql_path, "pg_hba.conf")
+
+        # pgsql user home must exist before we can initialize postgresql data
+        if not os.path.exists(self.pg_home):
+            os.mkdir(self.pg_home, mode=0o750)
+            shutil.chown(self.pg_home, user=self.pg_user, group=self.pg_group)
+
+        if not os.path.exists(asigra_postgresql_path):
+            os.mkdir(asigra_postgresql_path, mode=0o750)
+            shutil.chown(asigra_postgresql_path, user=self.pg_user, group=self.pg_group)
+
+        if not os.path.exists(asigra_pg_hba_conf):
+            proc = await Popen(
+                ['/usr/local/etc/rc.d/postgresql', 'oneinitdb'],
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE,
+                close_fds=True
+            )
+            output = (await proc.communicate())[0].decode()
+            if proc.returncode != 0:
+                self.logger.error(output)
+                self.logger.error('Failed to initialize postgresql:\n{}'.format(output))
+                return False
+
+            self.logger.debug(output)
+
+            await await Popen(
+                ['/usr/sbin/chown', '-R', '{}:{}'.format(self.pg_user, self.pg_group), asigra_postgresql_path],
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE,
+                close_fds=True
+            ).wait()
+
+            if not os.path.exists(asigra_postgresql_conf):
+                self.logger.error("{} doesn't exist!".format(asigra_postgresql_conf))
+                return
+
+            # listen_address = '*'
+            pg_conf_regex = re.compile('.*(#)\s{0,}?listen_addresses\s{0,}=\s{0,}([^\s]+)', re.M | re.S)
+            pg_conf_buf = []
+            rewrite = False
+
+            with open(asigra_postgresql_conf, "r") as f:
+                for line in f:
+                    if pg_conf_regex.search(line):
+                        def pg_re_replace(m):
+                            if m and len(m.groups()) > 1:
+                                return m.group(0)[m.end(1):m.start(2)] + "'*'" + m.group(0)[m.end(2):]
+
+                        line = pg_conf_regex.sub(pg_re_replace, line)
+                        rewrite = True
+                    pg_conf_buf.append(line)
+
+            if rewrite and pg_conf_buf:
+                with open(asigra_postgresql_conf, "w") as f:
+                    f.write("".join(pg_conf_buf))
+
+            if not os.path.exists(asigra_pg_hba_conf):
+                self.logger.error("{} doesn't exist!".format(asigra_pg_hba_conf))
+                return False
+
+            # allow local access
+            with open(asigra_pg_hba_conf, "a") as f:
+                f.write("host\tall\tall\t127.0.0.0/24\ttrust")
+
+        return True
+
+    # XXX
+    # We should probably write a postgresql middleware plugin
+    # We should also pass around a persistent DB handle
+    # XXX
+    async def asigra_database_exists(self):
```

Comment:
Given all the blocking operations (database) this should not be a coroutine, but a normal python method.